### ISSUE TITLE
Global library, queue and api

### ISSUE THREAD

**Issue by ThomasFrans** [hrkfdn/ncspot#1162]
I'm scared to ask this as almost anyone I've ever heard talk about development has told me to never use global variables. I might get send straight to hell for even proposing this change, but I'll do it anyway :stuck_out_tongue_closed_eyes: 

I wonder whether it would make sense to make `Library`, `Queue` and `Spotify` global variables. Over the last months, I've been learning Laravel and Angular and they both offer something like this, although not through raw global variables. Laravel has services that are accessible from anywhere in the program without needing a reference to anything. Angular has services that can be injected into components, basically acting similar to globally accessible state. Another example is the QT framework which keeps one global pointer to the `QApplication`. All of these make it way simpler to access often used logic/data from anywhere in the code.

This made me wonder whether it wouldn't make sense to do the same for `ncspot`. Currently a lot of views and other parts of the program pass references to a single instance of `Library`, `Queue` and `Spotify` around. If they were available as a global static reference (through a `OnceCell` or something similar), the code could become easier to read and write as the references wouldn't need to be passed around everywhere.

Another small advantage would be memory consumption and cpu usage improvements (although not by much). The references wouldn't need to be kept by everything and would also not have to be passed around anymore.

All of these (library, queue and api) are already safe to share between threads so that wouldn't require any changes.

I don't know about disadvantages of this approach. I've considered whether it would really be OK (from a code perspective) to make them global, but since there is only a single instance of each during the lifetime of ncspot, I think it makes sense.

It's just an idea. I'd love to get input on this. In case this would be considered/accepted, I would be willing to do the required changes and create a PR :slightly_smiling_face:

**Comment by hrkfdn** [comment]
Hey there! Don't be afraid to ask :)

I guess my question would we what the exact pain point is that should be resolved: is it preventing the annoyances of passing around all the references to objects? I thought about having a container structure that would contain all the cloneable `Library`, `Queue`, `Spotify` etc. references. It would still need to be passed around, but at least it's only one variable instead of 3-4.

**Comment by ThomasFrans** [comment]
Specifically, I was thinking in the same lines of the current async runtime implementation. There could be three static variables in the global scope, which could just be referred to like this.

```rust
use LIBRARY;

pub fn do_something(...) {
    ...
    LIBRARY.add_playlist(playlist);
}
```

There would be the advantage of not having to pass any reference to the library as its address can be accessed from anywhere, since it is a global variable.

Another advantage is that using the library from anywhere becomes way simpler since you don't need to get a reference to it from somewhere. You just include its name and start using it.

I also thought about the collection of library, spotify, queue... like you mentioned, but I don't think passing the references is the biggest problem. The real problem is getting the reference from somewhere to pass it in the first place, which would be solved by making it global.

Another problem it would solve is simplifying the synchronisation between UI and state. I've looked at the track deletion problem from a recent issue, and it appears to be caused by the playlist not being updated in the library after modifying the local copy of it. There is some logic to propagate changes from a playlist copy into the library, but that logic wouldn't be needed when everything has easy access to the library by default.

I know it's a big change and it's important to discuss this before making any decisions. In the end I would just want to simplify development. This is only one way I thought of. There might be other, better options. 🙂

**Comment by Bettehem** [comment]
I think your suggestion is pretty good and although it sounds like a lot of work, it could be a change worth making, as it would ease future development. I have had a problem like this in another rust project I've been working on, and this is a solution I might end up using there too!

**Comment by ThomasFrans** [comment]
As long as it is considered a good suggestion, I wouldn't mind to do the work. It's not very complicated work, just lots of it. I guess I would just fire up another ncspot, put on a good album and refactor away.

**Comment by hrkfdn** [comment]
Hmm, I'm also somewhat worried we're overlooking something because it _feels_ like a bad practice, but I also can't come up with anything concrete. I also totally see the upsides.

I wonder if you may run into issues because some of those instances require some initialization or depend on other objects, but I guess the compiler will tell you 😅 

Long story short: sounds good to me, happy to look at a PR if you come up with something :)

**Comment by ThomasFrans** [comment]
I started to work on this yesterday and I quickly realized why everyone says not to use global variables. When there would be a global library, queue and api, none of the method signatures can be trusted anymore. You could take a read lock on the queue, call a method and that method could take a write lock. I already mentioned this in the previous comment as well, but it is actually way worse with global variables, since with shared state using `Rc`/`Arc` the way it is implemented now, only certain methods can cause this problem (methods on structs/enums that hold a `Rc<Spotify>` for example). With globals, every method in the program can cause a deadlock, without a way to tell from their signature.

TLDR: I don't think global variables were a good approach after all. I even think it could make sense to do the opposite and remove some of the shared state in structs that don't need it to react to events, and make the methods take a reference instead as that more clearly conveys what the method does.

Sorry for wasting both of your time with this idea :frowning_face: . I now understand why globals aren't ever recommended.

**Comment by Bettehem** [comment]
Oh, don't worry. This sounded like a good idea if it would've worked nicely but using global variables in rust can be tricky indeed. You are probably correct that this isn't a good approach after all. Thanks looking at this though!

**Comment by hrkfdn** [comment]
Thanks for investigating! I'm not sure I fully understood the signature/locking problem: wouldn't it be the same since most of the nested locked objects are still behind an `Arc<RwLock<T>>`?

**Comment by ThomasFrans** [comment]
The nested objects are behind an `Arc<RwLock<T>>` but I think they can still cause problems. If there is code like this for example:

## With Globals

`playlist.rs`:
```rust
use LIBRARY;

...

impl Playlist {
    // This could be *any* function in ncspot, since they can all access the globals.
    // The caller has no idea about this function trying to aqcuire the lock.
    pub fn do_something(&self, ...) -> ... {
        LIBRARY.songs.read().unwrap();
        ...
    }
}
```

`main.rs`:
```rust
static LIBRARY: OnceCell<Library> = OnceCell::new();

fn main() {
    ...
    LIBRARY.songs.write().unwrap();
    playlist.do_something();    // do_something() doesn't take a library, so the caller wouldn't know it will try to access the LIBRARY.
// Since a global is accessible from any function, you would never be certain anymore that calling a function isn't going to lock (read/write) a Mutex/RwLock (of the global or inside the global) something that the caller is already holding a lock to.

}
```

## Without Globals and Passing References
The callee doesn't lock something that is also accessible from its caller, leaving potential deadlocks visible to the caller.

`playlist.rs`:
```rust
...

impl Playlist {
    pub fn do_something(&self, library: &mut Library, ...) -> ... {
        ...
    }
}
```

`main.rs`:
```rust

fn main() {
    ...
    let library = Library::new(...);
    library.songs.write().unwrap();
    playlist.do_something(LIBRARY.songs.write().unwrap());    // The caller is aware that do_something will write to LIBRARY and can see that a deadlock can occur

}
```

I haven't tested this code above, it's just to hopefully make clear where I think globals will cause a lot of problems with potential deadlocks. Everyone always says not to use globals but no one clearly explains *why* you shouldn't use them. I guess this is one of the reasons everyone informs you not to use them.
